### PR TITLE
Storing Zipped Schemas in Cluster Metadata

### PR DESCRIPTION
**Pull request by coderoshi**
This is a PR for #234, but this time it's against develop rather than master. It does what the name suggests. Creating a new schema will store the schema name and raw schema xml will first deflate the xml, and store the compressed data as cluster metadata. Getting the schema will inflate the xml and return it. The interface remains unchanged.

### CODE REVIEW

**Comment by hazen** [comment]
Mixed results:

```
Test Results:
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : fail
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : fail
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : fail
aae_test-bitcask             : fail
---------------------------------------------
4 Tests Failed
13 Tests Passed
That's 76.47058823529412% for those keeping score
```

Any expected failures?

**Comment by hazen** [comment]
Weird that YZ essential failed.  Might be my setup. Here's the output from that test:

```
================ yokozuna_essential failure stack trace =====================
{{badmatch,{error,enoent}},
 [{yokozuna_essential,read_schema,1,
                      [{file,"riak_test/yokozuna_essential.erl"},{line,238}]},
  {yokozuna_essential,setup_indexing,3,
                      [{file,"riak_test/yokozuna_essential.erl"},{line,253}]},
  {yokozuna_essential,confirm,0,
                      [{file,"riak_test/yokozuna_essential.erl"},{line,51}]}]}
=============================================================================
```

File: riak_test/yz_schema_admin.erl [hunk 1]
```diff
@@ -329,6 +331,18 @@ confirm_bad_schema(Cluster) ->
         end,
     yz_rt:wait_until(Cluster, F).
 
+%% @doc Confirm that creating a new shema gets added to the Ring
```
Comment (rzezeski) [inline riak_test/yz_schema_admin.erl]: Nitpick: It's not the ring metadata anymore, it's cluster metadata. Whole point was to get away from ring :).

File: src/yz_misc.erl [hunk 2]
```diff
@@ -262,6 +262,22 @@ set_ring_meta(Name, Default, Fun, Arg) ->
       fun set_ring_trans/2,
       {Name, Default, Fun, Arg}).
 
+-spec compress(iodata()) -> iolist().
+compress(Data) ->
+    Z = zlib:open(),
```
Comment (rzezeski) [inline src/yz_misc.erl:267]: Need to call close on `Z`.

File: src/yz_misc.erl [hunk 3]
```diff
@@ -262,6 +262,22 @@ set_ring_meta(Name, Default, Fun, Arg) ->
...
+    ok = zlib:deflateInit(Z),
+    Compressed = zlib:deflate(Z, Data, finish),
+    ok = zlib:deflateEnd(Z),
+    Compressed.
+
+-spec decompress(iodata()) -> iolist().
+decompress(Data) ->
+    Z = zlib:open(),
```
Comment (rzezeski) [inline src/yz_misc.erl:276]: Need to call close.

File: src/yz_schema.erl [hunk 4]
```diff
@@ -37,24 +37,28 @@ filename(SchemaName) ->
...
-            {ok, RawSchema}
+        {?YZ_DEFAULT_SCHEMA_NAME, undefined} ->
+            {ok, _} = file:read_file(?YZ_DEFAULT_SCHEMA_FILE);
+        {_, undefined} ->
+            {error, notfound};
+        {_, R} ->
+            {ok, iolist_to_binary(yz_misc:decompress(R))}
     end.
 
+-spec add_schema(schemas(), {schema_name(), compressed_schema()}) -> schemas().
```
Comment (rzezeski) [inline src/yz_schema.erl]: Am I missing something or is this not used anywhere?

File: include/yokozuna.hrl [hunk 5]
```diff
@@ -304,6 +305,8 @@
 -type raw_schema() :: binary().
 -type schema() :: xmerl_scan:document().
 -type schema_name() :: binary().
+-type compressed_schema() :: binary().
+-type schemas() :: orddict(schema_name(), compressed_schema()).
```
Comment (rzezeski) [inline include/yokozuna.hrl]: I believe this type is only use by `add_schema` which doesn't appear to be used anywhere.

**Comment by rzezeski** [comment]
@javajolt Your `enoent` failure is because it's failing to find the schema stored in the bench dir which is determined based on the `yz_dir` in your test config. Not sure if this PR was based on my recent change to `verify.sh`.

**Comment by hazen** [comment]
Thanks, @rzezeski. I asked @coderoshi earlier and he said that he did not have your latest `.riak_test.config` changes.  I was using this (instead of `riak_yz/deps/yokozuna`):

```
    {yz_dir, ["/tmp/yz-verify/yokozuna"]},
```

Interestingly, there does not seem to be this directory. Ugh.

**Comment by rzezeski** [comment]
Sigh, I forgot. The older verify.sh is broken. It will not create that dir anymore. Eric's tests probably passed because he didn't delete it. The easiest way is to just create a temporary branch where you merge develop into these changes. Really, any PR that is behind develop should be tested on a temp branch where develop is merged in because recent changes may be incompatible. Either that or the author should merge develop/rebase the PR.

**Comment by rzezeski** [comment]
Wait, scratch the last bit about verify.sh. This PR is only 4 commits behind. It has latest verify.sh which **does not** create a separate yokozuna dir anymore. You'll want `yz_dir` to point to the `riak_yz/deps/yokozuna` dir.

My comments on merging develop for testing still stand.

File: src/yz_schema.erl (hunk 4 above)
Comment (coderoshi) [inline src/yz_schema.erl]: Bah, it's unused. I originally populated cluster MD the via this function

**Comment by rzezeski** [comment]
- [x] code review
- [x] compile/unit
- [x] dialyze
- [x] riak tests

**Comment by rzezeski** [comment]
I ran into 2 test failures. Could also be caused by recent updates to riak or riak test. Need to investigate further.

```
Test Results:
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : pass
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : fail
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : fail
```

**Comment by rzezeski** [comment]
Okay, think I got this figured out. Eric, my guess is maybe you ran tests against the old code or something. Anyways, please take a look at my changes and +1 if you agree.

```
Test Results:
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : pass
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : pass
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : pass
---------------------------------------------
0 Tests Failed
17 Tests Passed
That's 100.0% for those keeping score
```

**Pull request merged**